### ISSUE TITLE
Value from mem address is not passed to read callback

### ISSUE THREAD

**Issue by VelpaChallenger** [TASEmulators/BizHawk#3813]
[//]: # "This description supports Markdown syntax. There's a cheatsheet here: https://guides.github.com/features/mastering-markdown/"
[//]: # "These lines are comments, for letting you know what you should be writing. You can delete them or leave them in."
[//]: # "Also, please don't waste your time writing until you've checked for similar Issues. Remember to check closed Issues too!"
[//]: # "One more thing: if you're on Linux, please comment on #1430 instead of opening an issue so we don't annoy the other devs."

### Summary
[//]: # "Briefly describe what's broken. Include relevant details: loaded core, loaded rom's hash, open tools, running scripts... You can embed a screenshot if it's easier to show the bug, but if you need more than one please put them at the end."
When running a lua script which registers a memory callback function to the on_bus_read hook, and the core calls the callback function, the value from the memory address is not passed.

So far I found it happens with BSNES and Genplus-gx, but it could potentially happen with other cores as well. Also with every game, for example Virtual Bart (Sega Genesis version, that is core Genplus-gx) and Contra 3: The Alien Wars (for BSNES). Both memory addresses are for the RNG. Contra 3: The Alien Wars' RNG starts advancing as soon as the Konami presentation starts (that is, before the title screen) and Virtual Bart's after the roulette starts spinning (start the game, skip the cutscene, then wait a few moments until the roulette starts spinning, you can also go to the practice area which makes the RNG advance as well).

### Repro
[//]: # "If you can't figure out the list of steps, delete this section and put 'heisenbug' in the summary somewhere. If a Lua script can cause the bug, you can embed that instead (as simple as possible please)."
1. Run Lua script which registers a memory callback function to on_bus_read hook (see below Lua script)
2. Go to section of the game which reads to the memory address passed to hook

Lua script:

```
function my_callback(addr, val, flags)
    print("addr being read is" .. string.format("%x", addr))
    print("value is" .. string.format("%x", val))
    print("flags is" .. flags)
    client.pause()
end
```

event.on_bus_read(my_callback, 0xFFFFC676) --FFFFC676 for Virtual Bart, E6 for Contra 3: The Alien Wars

### Output
[//]: # "Paste the contents of the error dialog if there is one (try Ctrl+C, it usually works), or paste the output from the Lua Console, or delete this section."
The value printed to the console is always 0.

### Host env.
[//]: # "List the computers you've found the bug with. If there's a version that doesn't have the bug, please put that in too. Here are some examples:"
- BizHawk 2.9.1 Win10 AMD

[//]: # "(screenshots, if applicable)"

[//]: # "That's it! If you'd like to help more, you could try a dev build (see Testing in the readme) or an older release. Click submit now and you can edit it later."

**Comment by Morilli** [comment]
I believe that is actually somewhat intentional, as reading data may have side-effects.

Also, the lua method documentation for `on_bus_read` states `Fires immediately before the given address is read by the core`... which makes me wonder, if the callback is to be called **BEFORE** reading the address (as is being done at least in the case of bsnes), how is the **read** address supposed to be passed to the callback?

**Comment by YoshiRulz** [comment]
Unlike writes, there's not a before and after value—unless you count registers but those are even easier to peek from the callback than memory is. So the value would be populated with the read function's return value as NesHawk does here, or equivalently from a peek (side-effects shouldn't come into this) at `addr` with the correct width.
https://github.com/TASEmulators/BizHawk/blob/b98942dcfdb7acbb39acc2864253d2bf55132fda/src/BizHawk.Emulation.Cores/Consoles/Nintendo/NES/NES.Core.cs#L1077-L1084
The other option is to fix the docs to reflect reality.

**Comment by VelpaChallenger** [comment]
Yes I didn't add it in the description, but it might be worth noting that the reason this behavior happens is because when the read memory callback function for each core (BSNES and Genplus-gx) is called, the value 0 is being passed as the value. My point being that as Yoshi says if the value is populated with the read function's return value, I think it should work, since only one read is happening anyway.

Here's what I mean:

https://github.com/TASEmulators/BizHawk/blob/6725363052eca77abda380c0491bd87ba836919e/src/BizHawk.Emulation.Cores/Consoles/Sega/gpgx64/GPGX.IDebuggable.cs#L70-L77

https://github.com/TASEmulators/BizHawk/blob/6725363052eca77abda380c0491bd87ba836919e/src/BizHawk.Emulation.Cores/Consoles/Nintendo/BSNES/BsnesCore.cs#L323-L329

As for the docs, I imagine it means before the address is read by the game's code, but not necessarily the core itself. So the way I understand it, whenever the game needs to read from an address, the read memory function is called (ReadMemory for NesHawk), so the game waits for the return value, meaning that until that value is not returned, the address was not read, even though of course it has to be read before by the core to be able to pass it to the game (and to the callback for that matter). I understood it that way the first time I read it, and if that's what it means, then yes it can be confusing, although I am not sure if it would require an update.

I am also giving it a try and seeing if I can do the fix myself. You can of course do a `memory.read_u16_be` or something of the sort inside your callback to read the value from the address, but I still think it would be nice to have the value returned and ready to be used inside the callback. I think it might also make it easier to understand why `val` is being passed, as otherwise it can cause confusion why val is being passed to the callback if it's always 0.

**Comment by Morilli** [comment]
> Unlike writes, there's not a before and after value

Well yes, but there's still a before and after state when considering side-effects, so I believe the distinction of before and after still matters. When the docs explicitly specify `before`, I implement `before`.

In that regard I'd definitely appreciate an update to the docs to clarify for both bizhawk devs and lua devs how exactly those functions should work. As it reads currently, calling the callback function before a value was read, but including the read value does not make sense.

**Comment by YoshiRulz** [comment]
It makes sense from the perspective of the script author: At the time the callback gets execution, the 'current'/'before' state is visible through the APIs, and since the 'future'/'after' state can't *easily* be reconstructed, the callback is provided those values via the parameters. It just happens that the `val` parameter is semantically the value of the memory at `addr` (and not a register), so for reads there is no difference between 'before' and 'after'.

Before in the callback API contract means before. If the implementation necessitates that the side-effects of a read must happen before or during a read, then the memory callback part will have to include an extra peek. (Or return `default`, which is fine but currently undocumented.)
